### PR TITLE
[SPARK-12992][SQL]: Update parquet reader to support more types when decoding to ColumnarBatch.

### PR DESCRIPTION
**Pull request by nongli**
This patch implements support for more types when doing the vectorized decode. There are
a few more types remaining but they should be very straightforward after this. This code
has a few copy and paste pieces but they are difficult to eliminate due to performance
considerations.

Specifically, this patch adds support for:
- String, Long, Byte types
- Dictionary encoding for those types.

### CODE REVIEW

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java [hunk 1]
```diff
@@ -180,6 +188,152 @@ public void readIntegers(int total, ColumnVector c, int rowId, int level,
...
+      }
+      rowId += n;
+      left -= n;
+      currentCount -= n;
+    }
+  }
+
+  @Override
+  public byte readByte() {
+    throw new UnsupportedOperationException("only readInts is valid.");
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: readInts -> readBytes

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java [hunk 2]
```diff
@@ -180,6 +188,152 @@ public void readIntegers(int total, ColumnVector c, int rowId, int level,
...
+    }
+  }
+
+
+  // This is used for decoding dictionary IDs (as opposed to definition levels).
+  @Override
+  public void readIntegers(int total, ColumnVector c, int rowId) {
+    int left = total;
+    while (left > 0) {
+    if (this.currentCount == 0) this.readNextGroup();
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: indent

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java [hunk 3]
```diff
@@ -180,6 +188,152 @@ public void readIntegers(int total, ColumnVector c, int rowId, int level,
...
+      int n = Math.min(left, this.currentCount);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: Should we have a defensive check to ensure that `currentCount > 0`?

Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: nvm, readNextGroup() can make sure that.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java [hunk 4]
```diff
@@ -52,15 +55,49 @@ public void skip(int n) {
...
+  @Override
+  public final long readLong() {
+    long v = Platform.getLong(buffer, offset);
+    offset += 8;
+    return v;
+  }
+
+  @Override
+  public final byte readByte() {
+    return (byte)readInteger();
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java]: Will a byte take 1 or 4 bytes?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java [hunk 5]
```diff
@@ -237,7 +233,8 @@ private void initializeInternal() throws IOException {
 
       // TODO: Be extremely cautious in what is supported. Expand this.
       if (originalTypes[i] != null && originalTypes[i] != OriginalType.DECIMAL &&
-          originalTypes[i] != OriginalType.UTF8 && originalTypes[i] != OriginalType.DATE) {
+          originalTypes[i] != OriginalType.UTF8 && originalTypes[i] != OriginalType.DATE &&
+          originalTypes[i] != OriginalType.INT_8 && originalTypes[i] != OriginalType.INT_16) {
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java]: Do we support ShortType (INT_16)?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java [hunk 6]
```diff
@@ -52,15 +55,49 @@ public void skip(int n) {
...
   @Override
-  public int readInteger() {
+  public final void readLongs(int total, ColumnVector c, int rowId) {
+    c.putLongsLittleEndian(rowId, total, buffer, offset - Platform.BYTE_ARRAY_OFFSET);
+    offset += 8 * total;
+  }
+
+  @Override
+  public final void readBytes(int total, ColumnVector c, int rowId) {
+    c.putBytes(rowId, total, buffer, offset - Platform.BYTE_ARRAY_OFFSET);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java]: ByteType should be INT32 in parquet, 4 bytes per SQL Byte

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java (hunk 1 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: This should be readInts. The only valid read* APIs that doesn't also decode definition levels is used to decode dictionary ids, which are always ints. I updated the comment for readIntegers() to try to capture this.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java (hunk 4 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java]: 4. Parquet only has 2 integers physical types: int32 and int64 and relies on encodings to compress bytes/shorts.

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Pull request closed without merging**